solutions/platinum/cses-1742.mdx

```diff
@@ -0,0 +1,254 @@
+		int mid = (l + r) >> 1;
+		return query(2 * node, l, mid, x, y) +
+		       query(2 * node + 1, mid + 1, r, x, y);
+	}
+};
+// EndCodeSnip
+
+struct point {
+	ll x, y;
+};
+struct segment {
+	ll a, b, x, y;
+};
+
+int n;
+ll max_y;
+vector<char> direction;
+vector<segment> segments, original_segments;
+map<char, pair<int, int>> dir;
+
+struct event {
+	ll x, y, id;
+	bool operator<(const event &other) {
+		if (x == other.x)
+			return (0 <= id && id < n && 0 <= other.id && other.id < n)
+			           ? segments[id].b < segments[other.id].b
+			           : id < other.id;
+
+		return x < other.x;
+	}
+};
+
+void coordinate_compression(vector<segment> &v) {
+	vector<ll> compressx, compressy;
+	compressx.push_back(0);
+	compressy.push_back(0);
+	for (auto seg : v) {
+		compressx.push_back(seg.a);
+		compressx.push_back(seg.x);
+		compressy.push_back(seg.b);
+		compressy.push_back(seg.y);
+	}
+	sort(all(compressx));
+	compressx.erase(unique(all(compressx)), compressx.end());
+	sort(all(compressy));
+	compressy.erase(unique(all(compressy)), compressy.end());
+	for (segment &seg : v) {
+		seg.a = lower_bound(all(compressx), seg.a) - compressx.begin() + 1;
+		seg.x = lower_bound(all(compressx), seg.x) - compressx.begin() + 1;
+
+		seg.b = lower_bound(all(compressy), seg.b) - compressy.begin() + 1;
+		seg.y = lower_bound(all(compressy), seg.y) - compressy.begin() + 1;
+
+		if (seg.a > seg.x) swap(seg.a, seg.x);
+		if (seg.b > seg.y) swap(seg.b, seg.y);
+		max_y = max(max_y, seg.b);
+		max_y = max(max_y, seg.y);
+	}
+}
+
+bool intersect(int idx) {
+	vector<event> events;
+	for (int i = 0; i <= idx; i++) {
+		if (segments[i].a == segments[i].x) {
+			events.push_back({segments[i].a, segments[i].b, i});
+		} else {
+			events.push_back({segments[i].a, segments[i].b, BEGINNING_VALUE});
+			events.push_back({segments[i].x, segments[i].y, ENDING_VALUE});
+		}
+	}
+	sort(all(events));
+
+	segtree sg(max_y);
+	segment last_vertical_line = {0, 0, 0, 0};
+	for (auto event : events) {
+		if (event.id == BEGINNING_VALUE) {
+			// Beginnig of a horizontal line
+			sg.update(1, 1, max_y, event.y, 1);
+			// Check for horizontal-horizontal intersection
+			if (sg.query(1, 1, max_y, event.y, event.y) > 1) return true;
+		} else if (event.id == ENDING_VALUE) {
+			// Ending of a horizontal line
+			sg.update(1, 1, max_y, event.y, -1);
+		} else {
+			// Checking for vertical-horizontal intersection
+			if (sg.query(1, 1, max_y, segments[event.id].b,
+			             segments[event.id].y) > 0)
+				return true;
+			// Checking for vertical-vertical intersection
+			if (segments[event.id].a == last_vertical_line.a &&
+			    segments[event.id].b <= last_vertical_line.y &&
+			    segments[event.id].y > last_vertical_line.b)
+				return true;
+			last_vertical_line = segments[event.id];
+		}
+	}
+	return false;
+}
+
+int main() {
+	scanf("%d ", &n);
+
+	dir['R'] = {1, 0};
+	dir['L'] = {-1, 0};
+	dir['U'] = {0, 1};
+	dir['D'] = {0, -1};
```

Comment:
initialize with this and make it a constant

Reply:
?